src/app/icd/client/DefaultICDClientStorage.cpp

```diff
@@ -488,9 +493,28 @@ CHIP_ERROR DefaultICDClientStorage::DeleteAllEntries(FabricIndex fabricIndex)
     return mpClientInfoStore->SyncDeleteKeyValue(DefaultStorageKeyAllocator::FabricICDClientInfoCounter(fabricIndex).KeyName());
 }
 
-CHIP_ERROR DefaultICDClientStorage::ProcessCheckInPayload(const ByteSpan & payload, ICDClientInfo & clientInfo)
+CHIP_ERROR DefaultICDClientStorage::ProcessCheckInPayload(const ByteSpan & payload, ICDClientInfo & clientInfo, bool & refreshKey)
 {
-    // TODO: Need to implement default decription code using CheckinMessage::ParseCheckinMessagePayload
+    uint32_t counter;
+    MutableByteSpan appData;
+    auto * iterator = IterateICDClientInfo();
+    while (iterator->Next(clientInfo))
+    {
+        if (CHIP_NO_ERROR ==
+            chip::Protocols::SecureChannel::CheckinMessage::ParseCheckinMessagePayload(clientInfo.shared_key, payload, counter,
+                                                                                       appData))
+        {
+            auto checkInCounter = (counter - clientInfo.start_icd_counter) % kCheckInCounterMax;
+            VerifyOrReturnError(checkInCounter > clientInfo.offset, CHIP_ERROR_INVALID_SIGNATURE);
+            clientInfo.offset = counter - clientInfo.start_icd_counter;
+            if (checkInCounter > kCheckInRolloverConstant)
+            {
+                refreshKey = true;
+            }
+            return CHIP_NO_ERROR;
+        }
+    }
+    iterator->Release();
     return CHIP_NO_ERROR;
```

Comment:
when we cannot find any key, we may still return the error, CHIP_ERROR_NOT_FOUND, then Check-In handler don't need to report the error to application.
In test, we can validate this error.

Reply:
@bzbarsky-apple thought?